src/storage/src/source/generator/clock.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use mz_ore::now::NowFn;
+use mz_repr::{Datum, Row};
+use mz_storage_types::sources::load_generator::{Event, Generator};
+use mz_storage_types::sources::MzOffset;
+use std::iter;
+use std::time::Duration;
+
+pub struct Clock {
+    pub interval: Duration,
+}
+
+impl Generator for Clock {
+    fn by_seed(
+        &self,
+        now: NowFn,
+        _seed: Option<u64>,
+        resume_offset: MzOffset,
+    ) -> Box<dyn Iterator<Item = (usize, Event<Option<MzOffset>, (Row, i64)>)>> {
+        let tick_ms: u64 = self.interval.as_millis().try_into().unwrap();
+        let mut last_offset = resume_offset.checked_sub(MzOffset::from(1));
+        Box::new(
+            iter::from_fn(move || {
+                let current_offset = MzOffset::from(now() / tick_ms);
+                let message = |tick: MzOffset, diff: i64| {
+                    let mz_now = mz_repr::Timestamp::new(tick.offset * tick_ms);
+                    let row = Row::pack_slice(&[Datum::MzTimestamp(mz_now)]);
+                    (0, Event::Message(current_offset, (row, diff)))
+                };
+                let progress = (0, Event::Progress(Some(current_offset + 1)));
+                let events = match last_offset {
+                    None => vec![message(current_offset, 1), progress],
+                    Some(last_offset) if last_offset == current_offset => vec![progress],
+                    Some(last_offset) => vec![
+                        message(last_offset, -1),
+                        message(current_offset, 1),
+                        progress,
+                    ],
+                };
+                last_offset = Some(current_offset);
```

Comment:
After staring at this for a while, I think there's a bug here if the system time ever goes backwards! I think something like this would be robust against backwards time jumps:

```diff
diff --git a/src/storage/src/source/generator/clock.rs b/src/storage/src/source/generator/clock.rs
index c11f0611c2..d139eef1ea 100644
--- a/src/storage/src/source/generator/clock.rs
+++ b/src/storage/src/source/generator/clock.rs
@@ -30,21 +30,29 @@ impl Generator for Clock {
         Box::new(
             iter::from_fn(move || {
                 let current_offset = MzOffset::from(now() / tick_ms);
+
+                // If the current offset is less than the last offset, the
+                // system time has gone backwards. However, load generator
+                // offsets are not allowed to regress. So emit a no-op progress
+                // message to stall. If and when the system time catches up,
+                // we'll start producing data.
+                if let Some(last_offset) = last_offset {
+                    if current_offset <= last_offset {
+                        return Some(vec![(0, Event::Progress(Some(last_offset + 1)))]);
+                    }
+                }
+
                 let message = |tick: MzOffset, diff: i64| {
                     let mz_now = mz_repr::Timestamp::new(tick.offset * tick_ms);
                     let row = Row::pack_slice(&[Datum::MzTimestamp(mz_now)]);
                     (0, Event::Message(current_offset, (row, diff)))
                 };
-                let progress = (0, Event::Progress(Some(current_offset + 1)));
-                let events = match last_offset {
-                    None => vec![message(current_offset, 1), progress],
-                    Some(last_offset) if last_offset == current_offset => vec![progress],
-                    Some(last_offset) => vec![
-                        message(last_offset, -1),
-                        message(current_offset, 1),
-                        progress,
-                    ],
-                };
+                let mut events = vec![];
+                if let Some(last_offset) = last_offset {
+                    events.push(message(last_offset, -1));
+                }
+                events.push(message(current_offset, 1));
+                events.push((0, Event::Progress(Some(current_offset + 1))));
                 last_offset = Some(current_offset);
                 Some(events)
             })
```

Reply:
I believe this is also fixed in the proposed alternative